### PR TITLE
Feat: ecs_compatiblity support + (optional) target

### LINKED ISSUE #35: ECS compatibility
The input isn't using a codec and generates events directly - setting the [`host` field](https://github.com/logstash-plugins/logstash-input-snmptrap/blob/v3.0.6/lib/logstash/inputs/snmptrap.rb#L83), which isn't compatible with ECS.

Plugin also [sets top-level fields](https://github.com/logstash-plugins/logstash-input-snmptrap/blob/v3.0.6/lib/logstash/inputs/snmptrap.rb#L85-L87) that depend on the SNMP payload, and thus might benefit from the introduction of a `target`.

*p.s. Alongside ECS `event_factory` support should be added*

### PR DESCRIPTION
**Pull request by kares**
The PR's goal is to avoid schema conflicts, since the plugin sets the `host` field, in a non ECS compatible way.

To aid plugin users to better isolate SNMP payload data the plugin supports an optional `target`.
Also, the plugin was not using Logstash's `recent event_factory` support.

### CODE REVIEW

File: docs/index.asciidoc [hunk 1]
```diff
@@ -23,14 +23,26 @@ include::{include_path}/plugin_header.asciidoc[]
 
 Read snmp trap messages as events
 
-Resulting `@message` looks like :
+Resulting `message` field looks like :
```
Comment (karenzone) [inline docs/index.asciidoc:26]:
```suggestion
Resulting `message` field resembles:
```

Comment (karenzone) [reply docs/index.asciidoc:26]: I know that this text was already there, but we can use this opportunity to tweak it. I try to use "resembles" because it won't be exact.  Not that I think anybody would *really* get confused over this.  :-D  But more precise language is always better.

**Review by karenzone — APPROVED** [review]
Docs review: One minor tweak for your consideration. Otherwise, LGTM!

File: docs/index.asciidoc [hunk 2]
```diff
@@ -58,6 +72,19 @@ input plugins.
 
 SNMP Community String to listen for.
 
+[id="plugins-{type}s-{plugin}-ecs_compatibility"]
+===== `ecs_compatibility`
+
+  * Value type is <<string,string>>
+  * Supported values are:
+    ** `disabled`: does not use ECS-compatible field names (fields might be set at the root of the event)
+    ** `v1`, `v8`: avoids field names that might conflict with Elastic Common Schema (for example, JMS specific properties)
```
Comment (yaauie) [inline docs/index.asciidoc:81]:
```suggestion
    ** `v1`, `v8`: avoids field names that might conflict with Elastic Common Schema (for example, SNMP specific properties)
```

File: lib/logstash/inputs/snmptrap.rb [hunk 3]
```diff
@@ -72,26 +92,34 @@ def build_trap_listener
...
+      rescue => e
+        @logger.error("Failed to create event", :exception => e, :backtrace => e.backtrace, :trap_object => trap)
       end
     end
-    @snmptrap.join
+    @snmptrap
   end # def snmptrap_listener
 
+  def process_trap(trap)
+    @logger.debug? && @logger.debug("SNMP Trap received: ", :trap_object => trap)
```
Comment (yaauie) [inline lib/logstash/inputs/snmptrap.rb:112]: previous code sent `Object#inspect` to ensure we logged a string. Is this change intentional?

```suggestion
    @logger.debug? && @logger.debug("SNMP Trap received: ", :trap_object => trap.inspect)
```

**Review by yaauie — COMMENTED** [review]
On track. I've left a doc suggestion, and one question about a change in logging format.

File: docs/index.asciidoc (hunk 2 above)
Comment (karenzone) [reply docs/index.asciidoc:81]: Good catch! Thank you

Comment (kares) [reply docs/index.asciidoc]: updated to https://github.com/logstash-plugins/logstash-input-snmptrap/pull/37/commits/faad777e738818895005e951680bc5fc48529fcc

File: spec/inputs/snmptrap_spec.rb [hunk 4]
```diff
@@ -1,13 +1,111 @@
...
+        trap # dummy manager - we're just using the Manager API to create traps
+      end
+    end
+  end
+
+  ecs_compatibility_matrix(:disabled, :v1, :v8) do |ecs_select|
+
+    let(:config) { super().merge 'ecs_compatibility' => ecs_compatibility }
+
+    context 'v1' do
```
Comment (yaauie) [inline spec/inputs/snmptrap_spec.rb:34]: Since ECS also has `v1` and friends, clearly indicating that this `v1` refers to the SNMP version will add clarity to the specs.

```suggestion
    context 'with an SNMP v1 trap' do
```

File: spec/inputs/snmptrap_spec.rb [hunk 5]
```diff
@@ -1,13 +1,111 @@
...
+        if ecs_select.active_mode == :disabled
+          expect( @event.get('host') ).to eql source_ip
+        else
+          expect( @event.get('host') ).to eql 'ip' => source_ip
+        end
+      end
+
+    end
+
+    context 'v2' do
```
Comment (yaauie) [inline spec/inputs/snmptrap_spec.rb:60]:
```suggestion
    context 'with an SNMP v2 trap' do
```

**Review by yaauie — APPROVED** [review]
Minor nit on naming spec contexts, otherwise LGTM.

Big :+1: to adding relevant and useful spec coverage.
